File: docs/src/main/java/jexamples/basics/pingpongdirect/Pinger.java
package jexamples.basics.pingpongdirect;

import se.sics.kompics.Kompics;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Positive;
import se.sics.kompics.Handler;
import se.sics.kompics.Start;

public class Pinger extends ComponentDefinition {

  Positive<PingPongPort> ppp = requires(PingPongPort.class);

  private long counter = 0;

  Handler<Start> startHandler =
      new Handler<Start>() {
        public void handle(Start event) {
          trigger(new Ping(), ppp);
        }
      };
  Handler<Pong> pongHandler =
      new Handler<Pong>() {
        public void handle(Pong event) {
          counter++;
          logger.info("Got Pong #{}!", counter);
          if (counter < 10) {
            trigger(new Ping(), ppp);
          } else {
            Kompics.asyncShutdown();
          }
        }
      };

  {
    subscribe(startHandler, control);
    subscribe(pongHandler, ppp);
  }
}

File: docs/src/main/scala/sexamples/basics/pingpongdirect/Pinger.scala
package sexamples.basics.pingpongdirect

import se.sics.kompics.sl._

class Pinger extends ComponentDefinition {

  val ppp = requires(PingPongPort)

  private var counter: Long = 0L;

  ctrl uponEvent {
    case _: Start => {
      trigger(Ping() -> ppp);
    }
  }

  ppp uponEvent {
    case Pong => {
      counter += 1L;
      log.info(s"Got Pong #${counter}!");
      if (counter < 10) {
        trigger(Ping() -> ppp);
      } else {
        Kompics.asyncShutdown();
      }
    }
  }
}
